Add covariance calculation

juliaschatz opened the pull request.

juliaschatz committed:
Add covariance calculation

JWCS commented on src/navx_node.cpp:
```diff
@@ -21,6 +21,12 @@
 #include "ahrs/AHRS.h"
 
 #define DEG_TO_RAD ((2.0 * 3.14159) / 360.0)
+#define GRAVITY 9.81 // m/s^2, if we actually go to the moon remember to change this
```
Since this is C++, should use `static const`'s instead, cleaner.

JWCS commented on src/navx_node.cpp:
```diff
@@ -52,9 +60,11 @@ int main(int argc, char** argv) {
   pnh->param<bool>("publish_euler", publish_euler, false);
   pnh->param<std::string>("device_path", device_path, "/dev/ttyACM0");
   pnh->param<std::string>("frame_id", frame_id, "imu_link");
+  pnh->param<int>("covar_samples", covar_samples, 100);
```
Should get changed to getParam();

JWCS replied:
Why are you down voting? If you didn't want to get stylistically and proceedurally better, with code review, why bother with a PR just to ignore it?

JWCS commented on src/navx_node.cpp:
```diff
@@ -52,9 +60,11 @@ int main(int argc, char** argv) {
   pnh->param<bool>("publish_euler", publish_euler, false);
```
Also, all of this should be at top-of-file documentation, where is it?

juliaschatz replied:
The documentation is in the readme and the launch file. It doesn't need to be in a third place.

JWCS replied:
This is THE most critical place to have the documentation! It's much more likely to be correct and up-to-date, nevermind inseperable from the code itself! 
When dealing with lots of nodes, and later needing to edit/work with them, not just interface with them, full in-file documentation is a must. 
The project is still small, but the most critical feature about ros is the documentation. You need to have good standards from the beginning; this isn't your code, it's the code of the people coming after you, yourself included.

JWCS commented on src/navx_node.cpp:
```diff
@@ -67,7 +77,62 @@ int main(int argc, char** argv) {
   ros::spin();
 }
 
+void calculate_covariance(boost::array<double, 9> &orientation_mat, boost::array<double, 9> &ang_vel_mat, boost::array<double, 9> &accel_mat) {
```
Should #include boost::array

JWCS commented on src/navx_node.cpp:
```diff
@@ -21,6 +21,12 @@
+  float ypr[3];
+  float ang_vel[3];
+  float accel[3];
+} Pose;
```
It would be better to name it something other than the generic "Pose": for example, there are already geometry_msgs msgs with that name as well.

JWCS commented on src/navx_node.cpp:
```diff
@@ -67,7 +77,62 @@ int main(int argc, char** argv) {
+  int count = std::min(seq-1, covar_samples);
+  if (count < 2) {
+    return;
+  }
```
I would have it return bool, whether it passed or not, and you can make the `2` be a tuning factor; 
That way you can choose not to publish the Imu msg until you have a covariance for it (should, ergo, take less than a sec, but no unclear data).

JWCS commented on src/navx_node.cpp:
```diff
@@ -67,7 +77,62 @@ int main(int argc, char** argv) {
+      accel_mat[idx] /= count - 1;
+    }
+  }
+
```
If you used an external source for algo/reference, would be good to include links, either here or at end of file. (I recognize, in this case, it is the definition, but comments are good)

JWCS commented on src/navx_node.cpp:
```diff
@@ -36,10 +42,12 @@ double frequency;
 // Global_Vars
 AHRS* com;
 int seq = 0;
+std::vector<Pose> poseHistory;
```
There needs to be a prototype definition of calculate_covariance up here. For super-long functions (I mean, holding y'all to a lower quality standard, this shouldn't happen too often naming-wise), this is a good form:
```
void calculate_covariance(
  boost::array<double, 9> &orientation_mat,
  boost::array<double, 9> &ang_vel_mat,
  boost::array<double, 9> &accel_mat
) {
```

JWCS reviewed (dismissed):
Looks, good, but *The documentation in the node is missing*. 
Otherwise, just cleanup, I don't see any big changes.

juliaschatz committed:
Minor cleanup changes

Rename Pose struct to OrientationEntry. Add calculate_covariance function prototype. Don't publish IMU message without a covariance.

JWCS commented on src/navx_node.cpp:
```diff
@@ -13,14 +13,31 @@
 
-#define DEG_TO_RAD ((2.0 * 3.14159) / 360.0)
+static const float DEG_TO_RAD = M_PI / 180;
+static const float GRAVITY = 9.81; // m/s^2, if we actually go to the moon remember to change this
```
Difference between floats (F32) and doubles (F64): when declaring a constant, F64's must have trailing .0, (ex 180.0), but F32's must also have trailing 'F', (ex 180.0F). 
Otherwise you're doing an implicit cast, which may get interpreted differently, especially in division, if not paying attention.

JWCS commented on src/navx_node.cpp:
```diff
@@ -13,14 +13,31 @@
+ */
+bool calculate_covariance(boost::array<double, 9> &orientation_mat, 
+                          boost::array<double, 9> &ang_vel_mat, 
+                          boost::array<double, 9> &accel_mat);
```
I don't particularly think, for internal function prototypes, the docs should go above the main, above the function declaration, as it clutters up the more useful information, if the docs are more than 1 line (1 line is solid). 
Style wise, it's easier to maintain / change / use if you avoid super long indents. Just put all of them below the function name: 
```
bool calculate_covariance(
    boost::array<..., 
    boost::array<...,
    boost::array<... /* Depending on coding standard, either this line will have the closing ');' or it will go on the next line, ex: */
);
```

JWCS commented on src/navx_node.cpp:
```diff
@@ -79,25 +158,21 @@ void update_callback(const ros::TimerEvent&) {
+
+  if (calculate_covariance(msg.orientation_covariance, 
+                           msg.angular_velocity_covariance, 
+                           msg.linear_acceleration_covariance)) {
```
See comment by function prototypes above. This can become more legible by moving the inputs down a line, and only indenting once more. 
Where you put the closing/opening braces/curly braces are a matter of company convention, I forget what google's is, I think they'll stay on last line of final argument. Can change for readability, especially for weirder things.

juliaschatz replied:
I find this style of long function calls/headers to be more readable.

JWCS replied:
I also haven't mentioned, b/c I haven't noticed, using github which doesn't show these things, but a (nearly universal) rule on column widths are 80 (soft) max; different companies, for particularly bad lines, will have a hard max of like 92 or something. 
Also, as you start getting more experienced, and you start learning how to use your tools better, you'll start "window"-ing your files, splitting portions of your screen more & more efficiently to have more reference files and terminal buffers available all on the same monitor. Code that nicely fits into 40-50 characters, or that hugs the left margin, becomes more efficient to work with and use. For slightly similar and completely disastrous results, google the Arrow Anti-pattern. 
You can leave this as is, it is not worth my own time to bring it up twice.

JWCS replied:
I unresolved this just so you could see the comment. It's resolved.

JWCS reviewed (changes requested):
Looks good, address remaining comments.

juliaschatz committed:
Minor style fixes

Explicitly use F32 values for constants. Move documentation into launch file.
Move function documentation to definition.

juliaschatz commented:
I'm happy to take suggestions about functional changes that should be made (eg, not publishing IMU message until we have a covariance) or important style fixes. I don't feel like it's worth my time to implement every tiny style change especially when we don't have a defined style guide and when I'm happy with my version.

JWCS commented on src/navx_node.cpp:
```diff
@@ -2,25 +2,50 @@
+/* Interface: 
+ *  Pub: 
+ *   imu/data (sensor_msgs/Imu)
+ *   imu/euler (geometry_msgs/Point)
```
You do want to specify a name for these publishers, because the topics are not always the same as what you intend it to be, and the types are often too generic. 
By 'name', I mean that the Publisher/Subscriber object, any boolean param to turn on/off, the callback, and any other related data should keep the same prefix. 
Ex: euler_pub, publish_euler, (not in this) euler_sub, euler_callback, euler_msg, euler_...
This also is important for when there are multiple pubs/subs of same type, in very nearby topics. 
For example, the map-global ekf would listen to "odometry/filtered" to publish an "odometry/filtered_map"
Example:
```
// Pub
//    imu_pub (sensor_msgs/Imu): "imu/data"
//    euler_pub (geometry_msgs/Point): "imu/euler"

// Sub
//    magnetometer_sub (...): "..."
```

JWCS commented on src/navx_node.cpp:
```diff
@@ -2,25 +2,50 @@
  * navx_node.cpp
  * Runs the Kauai Labs NavX, using modified NavX library
  * VERSION: 1.0.0
```
You forgot to bump version numbers for this PR. Usually, 1.X.X bump is for major (breaking or just big) changes, X.1.X is for normal changes, and X.X.1 is (for small projects like ours) documentation changes, or moving stuff around, with no "real" change to code, but bigger projects will use this for bugfixes as well, since there tend to be much more of them.

JWCS commented on src/navx_node.cpp:
```diff
@@ -2,25 +2,50 @@
+bool calculate_covariance(boost::array<double, 9> &orientation_mat, 
+                          boost::array<double, 9> &ang_vel_mat, 
+                          boost::array<double, 9> &accel_mat);
 
```
As you can see, this should all now go under Global Variables. If there's a lot of function prototypes, then I would add another comment and break them out, else leave them as a "variable". 
Custom types should go after Params & before Global Vars, b/c you can't have a param that's more complex than a basic type*, same with the other ros stuff defined before. 
The point is to keep the boilerplate in as consistent and priority-sorted order; even the functions after main() should go in (roughly) the order of complexity, because the smallest of helper functions will never need be changed nearly as much. 
(* other than vectors or maps, or a custom advanced ros type, but that's for config files, and never needed unless you are doing very unusual stuff; as for me, it's usually been for testing purposes, not normal).

JWCS commented:
There is a defined style guide, Google's C++ guide. Everything else I speak of from experience with working with long-running repos of code, infrequently but critically modified only a few times a year, with programmers of mixed experience. And misra C. 
Happiness and success are only as linked as pleasure and joy; one made instantly, the other enduring time; one easily produced, the other must be sought out. 
I am trying to teach, through these corrections, of paradigms, patterns, and styles that should help you; my personal opinion, while I appreciate your arguments to validate my reasoning, is that the result of all this time between us over these will make you a higher quality programmer. 
I was just gonna squash and merge this, but then I saw more than one thing that should get changed, so I'll leave this as the last review request.

juliaschatz committed:
Minor style changes

Rename variables. Bump version. Move declarations around.
